File: internal/trans/model/account.go
// GoToSocial
// Copyright (C) GoToSocial Authors admin@gotosocial.org
// SPDX-License-Identifier: AGPL-3.0-or-later
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

package trans

import (
	"crypto/rsa"
	"time"
)

// Account represents the minimum viable representation of an account for export/import.
type Account struct {
	Type                  Type            `json:"type" bun:"-"`
	ID                    string          `json:"id" bun:",nullzero"`
	CreatedAt             *time.Time      `json:"createdAt" bun:",nullzero"`
	Username              string          `json:"username" bun:",nullzero"`
	Domain                string          `json:"domain,omitempty" bun:",nullzero"`
	HeaderRemoteURL       string          `json:"headerRemoteURL,omitempty" bun:",nullzero"`
	AvatarRemoteURL       string          `json:"avatarRemoteURL,omitempty" bun:",nullzero"`
	DisplayName           string          `json:"displayName,omitempty" bun:",nullzero"`
	Note                  string          `json:"note,omitempty" bun:",nullzero"`
	NoteRaw               string          `json:"noteRaw,omitempty" bun:",nullzero"`
	Locked                *bool           `json:"locked"`
	Discoverable          *bool           `json:"discoverable"`
	URI                   string          `json:"uri" bun:",nullzero"`
	URL                   string          `json:"url" bun:",nullzero"`
	InboxURI              string          `json:"inboxURI" bun:",nullzero"`
	OutboxURI             string          `json:"outboxURI" bun:",nullzero"`
	FollowingURI          string          `json:"followingUri" bun:",nullzero"`
	FollowersURI          string          `json:"followersUri" bun:",nullzero"`
	FeaturedCollectionURI string          `json:"featuredCollectionUri" bun:",nullzero"`
	ActorType             int16           `json:"actorType" bun:",nullzero"`
	PrivateKey            *rsa.PrivateKey `json:"-" mapstructure:"-"`
	PrivateKeyString      string          `json:"privateKey,omitempty" mapstructure:"privateKey" bun:"-"`
	PublicKey             *rsa.PublicKey  `json:"-" mapstructure:"-"`
	PublicKeyString       string          `json:"publicKey,omitempty" mapstructure:"publicKey" bun:"-"`
	PublicKeyURI          string          `json:"publicKeyUri" bun:",nullzero"`
	SensitizedAt          *time.Time      `json:"sensitizedAt,omitempty" bun:",nullzero"`
	SilencedAt            *time.Time      `json:"silencedAt,omitempty" bun:",nullzero"`
	SuspendedAt           *time.Time      `json:"suspendedAt,omitempty" bun:",nullzero"`
	SuspensionOrigin      string          `json:"suspensionOrigin,omitempty" bun:",nullzero"`
}

type AccountSettings struct {
	AccountID string
}

File: web/source/settings/lib/types/account.ts
/*
	GoToSocial
	Copyright (C) GoToSocial Authors admin@gotosocial.org
	SPDX-License-Identifier: AGPL-3.0-or-later

	This program is free software: you can redistribute it and/or modify
	it under the terms of the GNU Affero General Public License as published by
	the Free Software Foundation, either version 3 of the License, or
	(at your option) any later version.

	This program is distributed in the hope that it will be useful,
	but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
	GNU Affero General Public License for more details.

	You should have received a copy of the GNU Affero General Public License
	along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

import { Links } from "parse-link-header";
import { CustomEmoji } from "./custom-emoji";

export interface AdminAccount {
	id: string,
	username: string,
	domain: string | null,
	created_at: string,
	email: string,
	ip: string | null,
	ips: [],
	locale: string,
	invite_request: string | null,
	role: any,
	confirmed: boolean,
	approved: boolean,
	disabled: boolean,
	silenced: boolean,
	suspended: boolean,
	created_by_application_id: string,
	account: Account,
}

export interface Account {
	id: string,
	username: string,
	acct: string,
	display_name: string,
	locked: boolean,
	discoverable: boolean,
	bot: boolean,
	created_at: string,
	note: string,
	url: string,
	avatar: string,
	avatar_static: string,
	header: string,
	header_static: string,
	followers_count: number,
	following_count: number,
	statuses_count: number,
	last_status_at: string,
	emojis: CustomEmoji[],
	fields: [],
	enable_rss: boolean,
	role: any,
	suspended?: boolean,
	source?: AccountSource;
}

export interface AccountSource {
	fields: any[];
	follow_requests_count: number;
	language: string;
	note: string;
	privacy: string;
	sensitive: boolean;
	status_content_type: string;
}

export interface SearchAccountParams {
	origin?: "local" | "remote",
	status?: "active" | "pending" | "disabled" | "silenced" | "suspended",
	permissions?: "staff",
	username?: string,
	display_name?: string,
	by_domain?: string,
	email?: string,
	ip?: string,
	max_id?: string,
	since_id?: string,
	min_id?: string,
	limit?: number,
}

export interface SearchAccountResp {
	accounts: AdminAccount[];
	links: Links | null;
}

export interface HandleSignupParams {
	id: string,
	approve_or_reject: "approve" | "reject",
	private_comment?: string,
	message?: string,
	send_email?: boolean,
}

export interface ActionAccountParams {
	id: string;
	action: "suspend";
	reason: string;
}

export interface AccountExportStats {
	media_storage: string;
	followers_count: number;
	following_count: number;
	statuses_count: number;
	lists_count: number;
	blocks_count: number;
	mutes_count: number;
}
